go/libraries/doltcore/merge/merge.go

```diff
@@ -230,6 +259,128 @@ func (merger *Merger) MergeTable(ctx context.Context, tblName string, opts edito
+	rootR, err := tbl.GetRowData(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	mergeR, err := mergeTbl.GetRowData(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	ancR, err := ancTbl.GetRowData(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	rootRP := durable.ProllyMapFromIndex(rootR)
+	mergeRP := durable.ProllyMapFromIndex(mergeR)
+	ancRP := durable.ProllyMapFromIndex(ancR)
+	collision := false
+	mergedRP, err := prolly.MergeMaps(ctx, rootRP, mergeRP, ancRP, func(left, right tree.Diff) (tree.Diff, bool) {
+		collision = true
+		return tree.Diff{}, false
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	if collision {
+		return nil, nil, errors.New("row collisions not supported yet")
+	}
+	updatedTbl, err := tableToUpdate.UpdateRows(ctx, durable.IndexFromProllyMap(mergedRP))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return updatedTbl, durable.IndexFromProllyMap(mergedRP), nil
+}
+
+// mergeProllySecondaryIndexes merges the secondary indexes of the given |tbl|,
+// |mergeTbl|, and |ancTbl|. It stores the merged indexes into |tableToUpdate|
+// and returns its updated value.
+func (merger *Merger) mergeProllySecondaryIndexes(ctx context.Context, postMergeSchema schema.Schema, mergedData durable.Index, tbl, mergeTbl, ancTbl, tableToUpdate *doltdb.Table) (*doltdb.Table, error) {
+	rootSet, err := tbl.GetIndexSet(ctx)
+	if err != nil {
+		return nil, err
+	}
+	mergeSet, err := mergeTbl.GetIndexSet(ctx)
+	if err != nil {
+		return nil, err
+	}
+	ancSet, err := ancTbl.GetIndexSet(ctx)
+	if err != nil {
+		return nil, err
+	}
+	mergedSet, err := mergeProllyIndexSets(ctx, merger.vrw, postMergeSchema, mergedData, rootSet, mergeSet, ancSet)
+	if err != nil {
+		return nil, err
+	}
+	updatedTbl, err := tableToUpdate.SetIndexSet(ctx, mergedSet)
+	if err != nil {
+		return nil, err
+	}
+	return updatedTbl, nil
+}
+
+// mergeProllyIndexSets merges the |root|, |merge|, and |anc| index sets based
+// on the provided |postMergeSchema|. It returns the merged index set.
+func mergeProllyIndexSets(ctx context.Context, vrw types.ValueReadWriter, postMergeSchema schema.Schema, mergedData durable.Index, root, merge, anc durable.IndexSet) (durable.IndexSet, error) {
+	mergedIndexSet := durable.NewIndexSet(ctx, vrw)
+
+	// Based on the indexes in the post merge schema, merge the appropriate
+	// ancestor, root, and merge indexes.
+	for _, index := range postMergeSchema.Indexes().AllIndexes() {
+		rootI, rootOK, err := root.GetIndex(ctx, postMergeSchema, index.Name())
+		if err != nil {
+			return nil, err
+		}
+		mergeI, mergeOK, err := merge.GetIndex(ctx, postMergeSchema, index.Name())
+		if err != nil {
+			return nil, err
+		}
+		ancI, ancOK, err := anc.GetIndex(ctx, postMergeSchema, index.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		mergedIndex, err := func() (durable.Index, error) {
+			// Build a new index based on |mergedData| if an index is missing.
```

Comment:
there is already logic to do this in `MergeTable`, do we need to duplicate it here?

Reply:
yes. In the new format, we are building this added index based on `mergedData`, which is the post-merge primary row data.

In the old format, we are building this added index based on the pre-merge data. Then we use the tblEditor's functionality to bring it up to date with the merged data.